### ISSUE TITLE
Different mass in cell with BURN_AWAY and BULK_DENSITY with HOLES

### ISSUE THREAD

**Issue by johodges** [firemodels/fds#13151]
**Describe the bug**
The attached case has different amounts of combustible mass allocated to cells when generated from a single OBST with a hole versus when generated with three separate OBSTs. The mesh without the HOLE will start to burn away much faster than the case with the hole. When I added the COMBUSTIBLE MASS output referenced in https://github.com/firemodels/fds/pull/13150, I saw the mass allocated to each grid cell was ~10x higher in the case with the HOLE than without the HOLE. The SURFACE DENSITY output value generally agrees with the input file but does not change over time which makes it difficult to figure out the root issue.

**To Reproduce**
[burnaway_test2.txt](https://github.com/user-attachments/files/16167685/burnaway_test2.txt)

**Comment by mcgratta** [comment]
```
&SLCF PBY=-0.6, QUANTITY='BULK DENSITY', CELL_CENTERED=T /
```
will show you that the bulk density in the obstruction with hole is 500 kg/m3 on the side and 312 kg/m3 on top.  This happens because the mass of the original block is passed on to the sub-blocks that are created when the `BURN_AWAY` attribute causes the original block to be broken up. I should be able to fix.

**Comment by mcgratta** [comment]
Try it and close if it fixes the problem.

**Comment by johodges** [comment]
Looks good on my test case and the more complicated case on which it was based. Thanks for jumping on it!

**Comment by drjfloyd** [comment]
@mcgratta 
Run this case and we get the correct mass from the obst (0.2^3 * 700 = 5.6)
Change the grid resolution to 15,15,15 and the mass jumps to 11.2.  If I revert back to OBT%MASS = OB%MASS/N_NEW_OBST it works.  


&HEAD CHID='bulk1'/

&MESH XB=0,2,0,2,0,2,IJK=20,20,20/

&TIME T_END=200/

&MISC Y_O2_INFTY=0.00001/

&SPEC ID='Pyrolysegas', FORMULA='CH1.7O0.7', SPECIFIC_HEAT=1.2/

&DUMP MASS_FILE=T/

&MATL ID='Papier',
      SPECIFIC_HEAT=1.4,
      CONDUCTIVITY=0.12,
      DENSITY=700.0,
      N_REACTIONS=1,
      HEAT_OF_REACTION=0,
      SPEC_ID(1,1)='Pyrolysegas',
      NU_SPEC(1,1)=1.0,
      A=1.0E+9,
      E=1.1E+5/

&SURF ID='Papier-Seite',
      RGB=146,8,77,
      HEAT_TRANSFER_COEFFICIENT=0.0,
      BURN_AWAY=.TRUE.,
      MATL_ID(1,1)='Papier',
      MATL_MASS_FRACTION(1,1)=1.0,
      THICKNESS(1)=0.15,
      EXTERNAL_FLUX=80/

&OBST ID='Obstruction', XB=0.8,1.0,1.1,1.3,1.5,1.7, BULK_DENSITY=700.0, SURF_ID='Papier-Seite'/

&BNDF QUANTITY='WALL TEMPERATURE'/
&BNDF QUANTITY='BURNING RATE'/

**Comment by mcgratta** [comment]
I'll take a look.

**Comment by mcgratta** [comment]
Does this change break the original case?

**Comment by drjfloyd** [comment]
I would assume so as that was what was there before. I'll double check.

**Comment by drjfloyd** [comment]
Yes. It does. I see what is happening though. OB%MASS is computed without knowledge of the HOLE through the OBST.  So when we go into the burnaway logic we take the original mass and divide over the much smaller number of single cells.  

The change you made worked in this specific case since the OBST was grid aligned. 

I think what we need is in READ_HOLE when we break up the obst, to recompute OB%MASS using the bulk density and the new sub OBST.

**Comment by drjfloyd** [comment]
testing a fix

**Comment by mcgratta** [comment]
I guess the big problem here is that when we break up a non-aligned `OBST`, the volume of the single cells is not going to equal the volume of the original obstruction. The original way of dividing the `OB%MASS` by the number of sub-cells worked, but the issue JH sent in obviously found a hole in that logic.

**Comment by drjfloyd** [comment]
My fix seems to work.  In READ_HOLE we divide the orginal OBST up into a bunch of temporary OBST in a manner that preserves the original volume. For each of those temp OBST I recompute the mass. The temp HOLE OBST gets tossed if it doesn't have a DEVC_ID or CTRL_ID. Now when we get to the subcell division for BURNWAY each sub OBST now has the correct mass for its volume and dividing by the number of subcells seems to be working.

**Comment by johodges** [comment]
Something I thought of while reading this. How are you handling HOLEs with a DEVC or CTRL in your new logic? Maybe a boolean to allow the user to choose whether the HOLE is considered as part of the burnaway mass? I am imagining the case where you want to prescribe a failure time if it does not occur, but allowing it to burn away until the prescribed failure time.

**Comment by mcgratta** [comment]
Let's at least let firebot test Jason's change. This stuff can be really tricky, so let's proceed with caution.

**Comment by drjfloyd** [comment]
My change did effect the current logic for how holes with DEVC or CTRL and BURN_AWAY are handled. It just ensured that the correct mass was assigned.  I looked at the current logic and I think it is OK.

If an OBST has burned away we don't bother checking the CTRL or DEVC and just force the OBST to be removed. It is only if there is mass left or the OBST isnt burnable that we check for the CTRL or DEVC.  So something like a room door that was burnable partially burned away and then was opened, that should work as any subOBST of the door that still had mass would still be checking the CTRL or DEVC status.

I can test this to make sure.

**Comment by drjfloyd** [comment]
I ran this case and burnaway with DEVC_ID seems to be working OK. The OBST starts to burnaway and at 120 s the remaining OBST is removed. The mass added to the domain is less than the total mass of the OBST and is consistent with the fraction of the OBST that burned away.

&HEAD CHID='bulk2'/

&MESH XB=0,2,0,2,0,2,IJK=15,15,15/

&TIME T_END=200/

&MISC Y_O2_INFTY=0.00001/

&SPEC ID='Pyrolysegas', FORMULA='CH1.7O0.7', SPECIFIC_HEAT=1.2/

&DUMP MASS_FILE=T/

&MATL ID='Papier',
      SPECIFIC_HEAT=1.4,
      CONDUCTIVITY=0.12,
      DENSITY=700.0,
      N_REACTIONS=1,
      HEAT_OF_REACTION=0,
      SPEC_ID(1,1)='Pyrolysegas',
      NU_SPEC(1,1)=1.0,
      A=1.0E+9,
      E=1.1E+5/

&SURF ID='Papier-Seite',
      RGB=146,8,77,
      HEAT_TRANSFER_COEFFICIENT=0.0,
      BURN_AWAY=.TRUE.,
      MATL_ID(1,1)='Papier',
      MATL_MASS_FRACTION(1,1)=1.0,
      THICKNESS(1)=0.15/

&SURF ID='HOT',TMP_FRONT=1500,COLOR='RED',HEAT_TRANSFER_COEFFICIENT=0/
&VENT PBZ=0,SURF_ID='HOT'/

&OBST ID='Obstruction', XB=0.8,1.0,1.1,1.3,1.5,1.7, BULK_DENSITY=700.0, SURF_ID='Papier-Seite',DEVC_ID='TIMER'/

&DEVC XYZ=0,0,0,QUANTITY='TIME',SETPOINT=120,ID='TIMER',INITIAL_STATE=T/

&SLCF QUANTITY='TEMPERATURE',PBY=1.9/
